src/rt/lifetime.d

```diff
@@ -281,15 +283,7 @@ bool __setArrayAllocLength(ref BlkInfo info, size_t newlength, bool isshared, co
         {
             if(isshared)
             {
-                try synchronized(typeid(ArrayAllocLengthLock))
-                {
-                    if(*length == cast(ubyte)oldlength)
-                        *length = cast(ubyte)newlength;
-                    else
-                        return false;
-                }
-                catch (Throwable t)
-                    assert(0, "Failed to synchronize.");
+                return cas(cast(shared)length, cast(ubyte)oldlength, cast(ubyte)newlength);
```

Comment:
Well, synchronized shouldn't leave a mutex locked after the pure function is done. But as far as this change goes, it has benefits besides purity. I would be concerned with making all synchronized calls impure, but that isn't this change.